### PR TITLE
[ROCKETMQ-224] add rocketmq client log4j2 logging support

### PR DESCRIPTION
**Pull request by lindzh**
When using RocketMQ client,we can only using logback or log4j for logging. If we using log4j2,there is no client log.

### CODE REVIEW

File: client/pom.xml [hunk 1]
```diff
@@ -37,13 +37,31 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-api</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-slf4j-impl</artifactId>
```
Comment (vongosling) [inline client/pom.xml]: Could we  put log4j2 test in rocketmq example module ?

File: pom.xml [hunk 2]
```diff
@@ -638,6 +638,16 @@
                 <artifactId>log4j-core</artifactId>
                 <version>2.7</version>
             </dependency>
+            <dependency>
+                <groupId>org.apache.logging.log4j</groupId>
```
Comment (vongosling) [inline pom.xml]: Why could i depend log4j2 core directly in dependency management?

File: client/pom.xml (hunk 1 above)
Comment (lindzh) [inline client/pom.xml]: As log4j is widely used in example,add this test in example will make it confused to choose log moudle.

File: client/src/main/java/org/apache/rocketmq/client/log/ClientLogger.java [hunk 3]
```diff
@@ -35,16 +36,19 @@
 
     private static Logger createLogger(final String loggerName) {
         String logConfigFilePath =
-            System.getProperty("rocketmq.client.log.configFile",
-                System.getenv("ROCKETMQ_CLIENT_LOG_CONFIGFILE"));
+                System.getProperty("rocketmq.client.log.configFile",
+                        System.getenv("ROCKETMQ_CLIENT_LOG_CONFIGFILE"));
```
Comment (zhouxinyu) [inline client/src/main/java/org/apache/rocketmq/client/log/ClientLogger.java]: Did you use the right code style? http://rocketmq.apache.org/docs/code-guidelines/

File: client/src/test/java/org/apache/rocketmq/client/log/ClientLogTest.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+import org.junit.Test;
+
+import java.util.Date;
+
+public class ClientLogTest {
+
+    @Test
+    public void testLog4j2() {
+        ClientLogger.getLog().info("logg4j test " + new Date());
```
Comment (zhouxinyu) [inline client/src/test/java/org/apache/rocketmq/client/log/ClientLogTest.java]: Any assert or thrown exception?

Comment (lindzh) [inline client/src/test/java/org/apache/rocketmq/client/log/ClientLogTest.java]: Assert has been added.

File: client/src/test/java/org/apache/rocketmq/client/log/ClientLogTest.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.rocketmq.client.log;
+
+import junit.framework.Assert;
```
Comment (lizhanhui) [inline client/src/test/java/org/apache/rocketmq/client/log/ClientLogTest.java]: We'd better use org.junit.Assert.

File: client/src/main/java/org/apache/rocketmq/client/log/ClientLogger.java [hunk 6]
```diff
@@ -27,16 +28,19 @@
     public static final String CLIENT_LOG_ROOT = "rocketmq.client.logRoot";
     public static final String CLIENT_LOG_MAXINDEX = "rocketmq.client.logFileMaxIndex";
     public static final String CLIENT_LOG_LEVEL = "rocketmq.client.logLevel";
+
     private static Logger log;
 
-    static {
-        log = createLogger(LoggerName.CLIENT_LOGGER_NAME);
+    public static Class logClass = null;
```
Comment (lizhanhui) [inline client/src/main/java/org/apache/rocketmq/client/log/ClientLogger.java]: Add logClass and its getter for test purpose only?

Comment (lindzh) [inline client/src/main/java/org/apache/rocketmq/client/log/ClientLogger.java]: Fixed it with reflect

File: client/src/test/java/org/apache/rocketmq/client/log/ClientLogTest.java [hunk 7]
```diff
@@ -32,15 +33,19 @@
         LOG_DIR = System.getProperty(CLIENT_LOG_ROOT, "${user.home}/logs/rocketmqlogs");
     }
 
+    // FIXME: Workarond for concret implementation for slf4j, is there any better solution for all slf4j implementations in one class ? 2017/8/1
```
Comment (vongosling) [inline client/src/test/java/org/apache/rocketmq/client/log/ClientLogTest.java]: we'd better remember this task and fix it later.

**Comment by vongosling** [comment]
Now, LGTM. this pr also fix the rocketmq's logger appender bug when using any concrete implementation, no matter log4j, log4j2 or logback.

File: client/src/main/java/org/apache/rocketmq/client/log/ClientLogger.java [hunk 8]
```diff
@@ -99,7 +111,12 @@ private static Logger createLogger(final String loggerName) {
     }
 
     public static Logger getLog() {
-        return log;
+        if (log == null) {
```
Comment (vsair) [inline client/src/main/java/org/apache/rocketmq/client/log/ClientLogger.java]: why not modify the init method?

**Pull request closed without merging**